### ISSUE TITLE
type mismatch when scaling a k3s cluster

### ISSUE THREAD

**Issue by larssb** [karmab/kcli#728]
We're operating part of our Kubernetes platform on VMWare/vSphere. [vApp options](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vm_admin.doc/GUID-A6F34BAC-BF8B-4513-AB8C-14891B439D2D.html) seems to be turned on for VM's deployed. This gives trouble when powering off and on a VM. When the VM is powered on again it gets irrelevant meta-data and therefore have no IP, a wrong hostname and the like.

The fix is easy, however manual and prolonged.

1. Mount the KCLI generated cloud-init ISO that it generated when the VM was deployed

Now the VM gets the correct IP, hostname and so forth.

---

To avoid the above we would love for `KCLI` to support flagging/configuring vApp options to be off for a deployed VM. Is this possible?

Thank you very much.

**Comment by karmab** [comment]
I dont think there is anything related to vapps in the vsphere provider

**Comment by larssb** [comment]
Aaah mayn ... can this maybe inspire you? https://github.com/vmware/pyvmomi/issues/894

Give it a look plz. Thank you very much.

**Comment by karmab** [comment]
it does!
addressed in https://github.com/karmab/kcli/commit/3c879580fa1fb530521db5bac366b1b1416157a9

**Comment by larssb** [comment]
Wicked! I'll ensure that we test this out soon!

**Comment by larssb** [comment]
Dang I get:

```text
Parameter numcpus doesn't have type <class 'int'>
```

When trying to scale in a new worker on a K3s cluster. 

Here's the `KCLI` command that we use to scale in the worker:

```text
kcli \
                    --debug \
                    --client "${__kcli_client}" \
                    scale kube k3s \
                    --paramfile "${__temp_file}" \
                    -P workers="${__scaleToWorkersCount}" \
```

The client is **vsphere**.

I've tried to debug some ... However, I have not found a way to fix this. Can you please give it a look.

KCLI is at version: `version: 99.0 commit: 3c87958 2024/10/18 Available Updates: False`

Thank you

**Comment by larssb** [comment]
So I debugged some more and I ended up going all the way back to: `version: 99.0 commit: eccf5a5 2024/06/13 Available Updates: True` before going all back to that version I went back, version jump by version jump to: `99.0.202408092118` before I got tired and went back to the one from 240613 that I was pretty sure would work.

In other words there's something that makes KCLI cause the `Parameter numcpus doesn't have type <class 'int'>` error between 240809 and 240613.

I hope this helps - at least a little bit.

Thank you.

**Comment by larssb** [comment]
In other news:

![image](https://github.com/user-attachments/assets/04d4997f-f5e4-4608-893f-292eb6602b63)

So adding `confspec.vAppConfigRemoved = True to the `vSphere` `__init__.py` file makes the `vApp option` become disabled. So that part is good.

**Comment by karmab** [comment]
your parameter file probably has numcpus defined as string, it should be int

**Comment by larssb** [comment]
We don’t use numcpus at all in the clusters kcli parameter files.I tried configuring numcpus and yes to an int. Made no difference.Further, it works up until the 13th of June this year. So something seems to have changed on this part after that date. --Mvh. / Kind regardsLars Bengtsson- Health is wealth, movement is medicine -On 18 Oct 2024, at 20.26, Karim Boumedhel ***@***.***> wrote:﻿
your parameter file probably has numcpus defined as string, it should be int

**Comment by karmab** [comment]
your example is wrong, it should be numcpus and not num_cpus, please try with that

**Comment by larssb** [comment]
I'll try tomorrow with `numcpus` and not `num_cpus`. I don't know how I was able to do that glaring mistake. However, I still don't understand why it stopped working. We haven't had to specify `numcpus` for as long as I remember. Also we can't really use `numcpus` ... as we wish to differ the `vCPU` count between control-plane and worker nodes.

**Comment by larssb** [comment]
Here's the result of running on `kcli info keyword numcpus` on `KCLI` version: _5f7a20e 2024/10/20_

```text
Providing information about keyword numcpus...
Default value: 2
Current value: 2
Detailed information:
Number of vcpus. Allow overcommitting
```

Also setting `numcpus` and **NOT** `num_cpus` ;-) in the `KCLI parameter file` does not make any difference. Still getting:

```text
Parameter memory doesn't have type <class 'int'>
```

Tried to debug further to pinpoint the version of `KCLI` where this happens.

- pinned `KCLI` to the newest version on 240630. Result: **WORKS**
- on and on until I found the version of `KCLI` where this breaks. The version is: `0eb9445 2024/07/12` and the `git commit` `https://github.com/karmab/kcli/commit/3546259bcda66d4a1ae38022fed7f35091749fc3`

In this version: `0eb9445 2024/07/12` I see this error: `Parameter sdn doesnt have type <class 'str'>`

And that's fair because we do: `sdn: ` and this is the comment we have on that parameter:

```text
# sdn needs to be "None". In other words have no value
# for us specifically. As we don't want the default K3s CNI
# flannel. And because we're handling the CNI ourselves.
# By using Cilium
sdn:
```

We've discussed this parameter being `NULL` way back in time. We need to do this in order to signal to `KCLI` to **NOT** do anything on `Cilium`.

Looking at the code from `https://github.com/karmab/kcli/commit/3546259bcda66d4a1ae38022fed7f35091749fc3`:

```python
                    correct_type = type(parameters[parameter])
                    if type(overrides[parameter]) != correct_type:
                        error(f"Parameter {parameter} doesnt have type {correct_type}")
                        sys.exit(1)
```

This part seems to be the most interesting part. So in other to e.g. support `sdn` being _null_. There would be a conditional on e.g.: `if ! provider = k3s` ... e.g. ... However, this would be annoying for contributors and maintainers of the `kcli` code.

What do we do here @karmab ?

Thank you

**Comment by karmab** [comment]
after your numcpus change, you have
Parameter memory doesn't have type <class 'int'>

so it means you have the same kind issue with memory beeing defined as string somewhere, set it explicitely in your parameter file just to confirm this is indeed the issue

**Comment by karmab** [comment]
I've improved logging via https://github.com/karmab/kcli/commit/113e20026ecf4e82f679774011f0292c8cc98884 so please use that so that we can see what value (and type) is being provided

**Comment by larssb** [comment]
Yes setting `memory` explicitly makes it work together with having `numcpus` set as well in the `KCLI parameter file`. However, as I write in my https://github.com/karmab/kcli/issues/728#issuecomment-2425218107 comment. We want to use `ctlplane_numcpus`, `worker_numcpus` and so on. To differ the resource setup between the node types.

Is this still possible with the change that happened in https://github.com/karmab/kcli/commit/3546259bcda66d4a1ae38022fed7f35091749fc3 ???

Thank you

**Comment by karmab** [comment]
yes, definitely, dont set it and use latest so that we can see which values you are passing

**Comment by larssb** [comment]
Don't set `ctlplane_numcpus, worker_numcpus, ctlplane_memory and worker_memory`? Are those the values that you refer to - when you say `don't set it`?

**Comment by larssb** [comment]
Looking at what you committed in [print wrong type/value when mismatch
](https://github.com/karmab/kcli/commit/113e20026ecf4e82f679774011f0292c8cc98884) - I think you just fixed it.

This line ( in kvirt/baseconfig.py ): `if parameter_value is not None and override_type != parameter_type:` has the effect that we no longer actually have to declare `numcpus` and `memory`. As they can now ( again ), as before https://github.com/karmab/kcli/commit/3546259bcda66d4a1ae38022fed7f35091749fc3 came in, be `None`.

And because the default values of `worker_numcpus`, `worker_memory` and `ctlplane_numcpus`, `ctlplane_memory` are default `None` in `https://github.com/karmab/kcli/blob/main/kvirt/cluster/k3s/kcli_default.yml` they also slip through as allowed because of these lines:

```text
if parameter_type == str and override_type == type(None):
    continue
```

( the lines above haven't changed for 3 months ).

So I think I now know why I was confused on our call and that we can conclude that you fixed it via [113e200](https://github.com/karmab/kcli/commit/113e20026ecf4e82f679774011f0292c8cc98884).

So I think there's nothing more to do here - do you agree @karmab?

**Comment by larssb** [comment]
:-D
